packages/plugins/plugin-device-token/src/DeviceTokenPlugin.tsx

```diff
@@ -0,0 +1,62 @@
+import {
+  PlatformPlugin,
+  SegmentClient,
+  PluginType,
+} from '@segment/analytics-react-native';
+import { Platform } from 'react-native';
+import messaging from '@react-native-firebase/messaging';
+
+export class DeviceTokenPlugin extends PlatformPlugin {
+  type = PluginType.enrichment;
+
+  async configure(analytics: SegmentClient) {
+    const isAuthorized = await this.checkUserPermission();
+    this.analytics = analytics;
+
+    if (isAuthorized) {
+      this.retrieveDeviceToken();
+    } else {
+      this.analytics?.logger.warn('Not authorized to retrieve device token');
+    }
+  }
+
+  async retrieveDeviceToken() {
+    if (Platform.OS === 'ios') {
+      let APNSToken = await messaging().getAPNSToken();
+      if (APNSToken !== null) {
+        this.setDeviceToken(APNSToken);
+      }
```

Comment:
You can go this way but there's even an extra way you can simplify this. I would actually have done it "backwards" by making the "reusable" function not setting the device token but instead one for getting the device token value in a standard way:

```
private async function getDeviceToken(): string | undefined {
  if (Platform.OS == 'ios') {
    return await messaging().getAPNSToken() ?? undefined; // This way we standard nulls and undefined into undefined only
  }
  if (Platform.OS == 'android') {
    let deviceToken = await messaging().getToken() ?? undefined;
    if (deviceToken !== undefined && deviceToken.length > 0) {
      return deviceToken;
    } else {
      return undefined;
    }
  }
  logger.warn("not available");
  return undefined;
}
```

This has the advantage that now your platform specific code is just one pure function without side effects. In theory you could just unit test this function for platform code and mock this one so that you don't have to mock platform on the rest.

Second advantage is that the output values now are normalized regardless of platform so now your `retrieveDeviceToken()` only needs to call this one and function normally:

```
let token = await getDeviceToken();
if (token !== undefined) {
  await context.set({ device: { token: token }});
  // etc
}
```

Reply:
ohhh that makes a lot of sense. I updated it a bit differently but I think it works!